### ISSUE TITLE
OpenGL-related 3D code broken after merge

### ISSUE THREAD

**Issue by stohrendorf** [opentomb/OpenTomb#117]
After the first merge the opengl code got broken. I guess I have misaligned the data structures and accidentally changed code flow somewhere.

**Comment by cochrane** [comment]
This is currently blocked by #114 (or at least I'm not going to do any work on it until that is resolved). Though that one should, hopefully, be completed soon.

**Comment by stohrendorf** [comment]
Either the cam pos is wrong or the geometry (though I think it is the latter case); after printing some debug infos, I discovered e.g. that `cam_pos=(29000,27666,900)` whereas the bounding box of the respecting room mesh is `bbMin=1024/1024/-2304` and `bbMax=13312/15360/3840`.

**Comment by vvs-** [comment]
The camera position is unreliable. There are some bugs where it ends up in a wrong room.

**Comment by vvs-** [comment]
Camera is definitely not working. Either using keyboard or mouse its position isn't changing. But the `goto` command seems to update it, though nothing is visible. The video memory is exhausted but disabling player entity helps with that instantly.

**Comment by vvs-** [comment]
The video memory was eaten by createTextures function.

**Comment by vvs-** [comment]
Reverting to old version of bordered_texture_atlas completely fixed memory problem but not the rendering :disappointed:

**Comment by stohrendorf** [comment]
Damn... I will look into the diffs for the atlas.

**Comment by stohrendorf** [comment]
Have you also reverted the 2D BSP Tree?

**Comment by vvs-** [comment]
Of course, it couldn't be done without it.

**Comment by stohrendorf** [comment]
Then I think that's the broken code, as the diffs for the texture atlas was pretty straight forward and included no critical changes, except for the tree.

**Comment by vvs-** [comment]
Yes, I looked at them too and found nothing suspicious. Then I resorted to drastic measures and reverted the whole thing.

**Comment by stohrendorf** [comment]
OK, I'm taking a look at it.

**Comment by stohrendorf** [comment]
Interestingly, I'm seeing a few "sparks" in TR3, which means that *something* is there.

**Comment by stohrendorf** [comment]
Also, now that I can hear FX, when I move slightly to the left in TR3 *Jungle*, I hear the scream of death. I guess Lara is simply placed somewhere in the nimbus.

**Comment by vvs-** [comment]
I focused on TR1, but now that you mentioned it, isn't a sound pitch too high? I'm not sure.

**Comment by stohrendorf** [comment]
It is, I think it's playing 22.05k sound at 44.1k. Somewhere it's missing an upsampling, but I haven't yet taken a deeper look into it, as it isn't the main problem right now.

**Comment by vvs-** [comment]
Yes, but it's still a bug so there should be an issue for it.

**Comment by vvs-** [comment]
The initial position is correct but she's dead at that instant. Also, it doesn't matter as it should render Lara anywhere in noclip mode.

**Comment by vvs-** [comment]
The problem is that it's not only rendering is broken but other things as well. Collision logic doesn't work properly and Lara now can die even in noclip state. And she moves in that state too.

**Comment by stohrendorf** [comment]
I have also discovered that the move direction when jumping is completely non-deterministic.

**Comment by vvs-** [comment]
> Interestingly, I'm seeing a few "sparks" in TR3, which means that something is there.

It is! I've spawned passport (model 71) on the same position as Lara is and it produced visible animation. But as soon as camera is moved away it's disappeared.

**Comment by stohrendorf** [comment]
Maybe it's a clipping problem.

**Comment by vvs-** [comment]
I thought about it. As so many things are broken simultaneously, I see no choice but disable everything and enable it step by step until something would work. That's a tedious work.

**Comment by vvs-** [comment]
The strange thing is that there is nothing visible at all if the debug information is off. I'm curious what's the difference.

**Comment by vvs-** [comment]
The passport animation now shows collision mesh with r_coll.

**Comment by vvs-** [comment]
It definitely looks like a frustum problem.

**Comment by stohrendorf** [comment]
I want to point to pull request #130, because that fixes some problems and I just made it able to build on windows again, yet I know this is another huge diff.

**Comment by vvs-** [comment]
I'll test it tomorrow. Then we'll have some confidence in it.

**Comment by stohrendorf** [comment]
I'm not sure, but the vertex positions stored in the VBOs seem a bit strange. Here's a representative excerpt: 
```
X, Y, Z
0.18859863, 1, 0.013427734
0.52587891, -514, -102
-511, 0, 0.70676112
0, 0.70745224, 0
0.18859863, 0.18859863, 0.18859863
1, 0.017089844, 0.52587891
-209, -69, -511
0, -0.70710677, 0
0.70710677, 0, 0.84790039
0.84790039, 0.84790039, 1
```

**Comment by cochrane** [comment]
Note that this is not just positions, but also normal vectors and colors. The values do seem weird, though. Still no time to look into it in detail; could you tell me what file creates the values you're seeing?

**Comment by cochrane** [comment]
I think there was some confusion about ownership rules among graphics objects, which led to lots of use of shared_ptr and so on in places where it provided no benefit (e.g. Shader manager). I wrote a short document on the wiki explaining what the rules were like prior to refactoring: https://github.com/opentomb/OpenTomb/wiki/Ownership-rules-for-graphics-objects please update to document other such rules; I think this can help resolve issues like this.

**Comment by stohrendorf** [comment]
It's `JUNGLE.TR2` and `mesh.cpp`; by the way, after putting some debug output into the renderer, I saw that the mesh vertices actually get sent to the graphics card.

**Comment by vvs-** [comment]
After loading a level m_rListActiveCount in Render::DrawList is zero, while r_list_active_count was 30 in the old version. That's for *Caves* level.

**Comment by vvs-** [comment]
I don't know why Lara model isn't rendered yet. But the rooms can't be rendered just because they're not in render list.

**Comment by stohrendorf** [comment]
I have hacked around in the luastate branch (the render list is now an actual render list, meaning everything in there will be rendered, and the rooms don't have a `is_in_r_list` flag), and managed to dump the vertex positions of all rendered vertices in `Render::renderMesh` after applying the used transformation matrix. Here's the outcome: http://earvillage.net/opentomb/vertex-dump.png

As you can see, data is actually sent to the graphics card, but it seems incomplete.

**Comment by vvs-** [comment]
Looking at OpenGL call traces I don't see anything unusual. There are geometry buffers pushed to GPU and they look exactly the same as that on a working version. I can't see what's inside but I'm starting to suspect that it's a GUI problem. Can't it be that GUI overlay is hiding the output?

**Comment by vvs-** [comment]
Was GUI code fully moved to shader pipeline? I remember that it was planned but have no details.

**Comment by stohrendorf** [comment]
I just discovered that the shader stages are destroyed after linking the programs, resulting in using invalid shader programs. I'm currently hacking at this.

**Comment by vvs-** [comment]
I see that compiled programs are deleted after linking. But that always was the case. I guess they are not needed after that.

**Comment by vvs-** [comment]
I don't see any other difference except texture coordinates and broken polygon visibility check.

**Comment by stohrendorf** [comment]
Yet I had a bunch of error logs saying that an invalid shader program was used. After persisting the shader stages they disappeared. Additionally, I just found out that the model/view/projection matrices are deforming the vertices in a way that they are just out of sight. On the left you see the undeformed mesh, on the right after the shader, including the white camera frustum. The "room" is completely malformed and out of view.

![shader](https://cloud.githubusercontent.com/assets/7853507/8565260/38b6c34c-2554-11e5-9075-eb547528d1b1.png)

**Comment by vvs-** [comment]
That's possible. I didn't look at matrices at all.

**Comment by vvs-** [comment]
Also, some rooms should be out of view because they are overlapping. There is a special code in frustum manager that make the geometry invisible.

**Comment by stohrendorf** [comment]
That doesn't explain why the geometry is flattened.

**Comment by vvs-** [comment]
I don't know. The problem is that only TeslaRus knows all the code and Cochrane knows the rendering code which he wrote. The code is very complex because the original Core engine actually used 4D for their geometry.

**Comment by vvs-** [comment]
I wouldn't dare to replace the original code unless there is no other choice left. We can't even properly test it until rendering is restored. We are risking to introduce even more new bugs. Right now I see the most probable candidates in the issues which we already know. I believe that the safe way is to restore Lara model rendering and then go step by step.

**Comment by stohrendorf** [comment]
Well, after switching all those matrix and vector manipulations to bullet, I will go to bed now after a long debugging night, and then start to investigate the math. It shouldn't be that hard, I studied it :+1:

**Comment by vvs-** [comment]
The simple refactoring shouldn't require changing the algorithm which is already proven and working. The bugs can't be that complex and the fix should be as simple as just finding not intended changes and restoring original behavior of the code.

**Comment by cochrane** [comment]
Gui code is all shader based, yes. Also it is perfectly legal to delete shader stages after attaching to a program (they use a form of reference counting internally). If you have questions about how something works, please, please, please ask first and hack later.

**Comment by stohrendorf** [comment]
I'm curious: are vertex shaders generally used to apply the projection matrix? I thought that projection was one of the very last steps of rendering.

**Comment by cochrane** [comment]
Found the problem, about to fix it. For the future, please remember: btTransform is only meant and suitable for affine transforms, not arbitrary 4x4 matrices. In fact it doesn't store the bottom row at all. This means it cannot be used for projection matrices, and anything that has the projection matrix included.

**Comment by stohrendorf** [comment]
Yes I'm totally aware of that. See also #140 if you're going to fix it, I'll keep my hands off for a while and wait.

**Comment by vvs-** [comment]
I think this should be closed now. We have other issues for all known bugs.

**Comment by Lwmte** [comment]
Only bug which is left is broken alpha test, which has separate ticket (#158), so I'm closing this.
